lib/pf/pki_provider/inverse.pm

```diff
@@ -0,0 +1,141 @@
+package pf::pki_provider::inverse;
+
+=head1 NAME
+
+pf::pki_provider::inverse
+
+=cut
+
+=head1 DESCRIPTION
+
+pf::pki_provider::inverse
+
+=cut
+
+use strict;
+use warnings;
+use Moo;
+use WWW::Curl::Easy;
+use URI::Escape::XS qw(uri_escape uri_unescape);
+
+use pf::log;
+
+=head2 uri
+
+=cut
+
+has uri => ( is => 'rw' );
+
+=head2 username
+
+=cut
+
+has username => ( is => 'rw' );
+
+=head2 password
+
+=cut
+
+has password => ( is => 'rw' );
+
+=head2 profile
+
+=cut
+
+has profile => ( is => 'rw' );
+
+=head2 country
+
+=cut
+
+has country => ( is => 'rw' );
+
+=head2 state
+
+=cut
+
+has state => ( is => 'rw' );
+
+=head2 organisation
+
+=cut
+
+has organisation => ( is => 'rw' );
+
+=head2 get_cert
+
+=cut
+
+sub get_cert {
+    my ($self,$args) = @_;
+    my $logger = get_logger();
+    my $uri = $self->uri;
+    my $username = $self->username;
+    my $password = $self->password;
+    my $email = $args->{'certificate_email'};
+    my $dot1x_username = $args->{'certificate_cn'};
+    my $organisation = $self->organisation;
+    my $state = $self->state;
+    my $profile = $self->profile;
+    my $country = $self->country;
+    my $certpwd = $args->{'certificate_pwd'};
+    my $curl = WWW::Curl::Easy->new;
+    my $request =
+        "username=" . uri_escape($username)
+      . "&password=" . uri_escape($password)
+      . "&cn=" . uri_escape($dot1x_username)
+      . "&mail=" . uri_escape($email)
+      . "&organisation=" . uri_escape($organisation)
+      . "&st=" . uri_escape($state)
+      . "&country=" . uri_escape($country)
+      . "&profile=" . uri_escape($profile)
+      . "&pwd=" . uri_escape($certpwd);
+    my $response_body = '';
+    $curl->setopt(CURLOPT_POSTFIELDSIZE,length($request));
+    $curl->setopt(CURLOPT_POSTFIELDS, $request);
+    $curl->setopt(CURLOPT_WRITEDATA, \$response_body);
+    $curl->setopt(CURLOPT_HEADER, 0);
+    $curl->setopt(CURLOPT_DNS_USE_GLOBAL_CACHE, 0);
+    $curl->setopt(CURLOPT_NOSIGNAL, 1);
+    $curl->setopt(CURLOPT_URL, $uri);
+
+    # Starts the actual request
+    my $curl_return_code = $curl->perform;
+
+    if ($curl_return_code == 0) {
```

Comment:
You should check that AND the HTTP return code. Because this could be a 401 or 403 or 500. Then it would be failing silently and inserting crap in the rest of the application